docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,339 @@
+   
+   Possible causes:
+
+   - You have retried for longer than the deduplication duration, but never got a meaningful answer, e.g., because the timeout (gRPC deadline) is too short.
+     For example, this can happen due to long-running Daml interpretation when using the :ref:`Command Service <command-service>`.
+
+   - The application clock drifts significantly from the participant's or ledger's clock.
+
+   - There are unexpected network delays.
+
+   - Submissions are retried internally in the participant or Daml ledger and those retries do not stop before ``B`` is over.
+
+#. Unacceptable changes cause infinite retries
+
+   You need business workflow knowledge to decide that retrying does not make sense any more.
+   Of course, you can always stop retrying and accept that you do not know the outcome for sure.
+
+
+Unknown processing time bounds
+==============================
+
+Finding a good bound ``B`` on the processing time is hard, and there may still be unforeseen circumstances that delay processing beyond the chosen bound ``B``.
+You can avoid these problems by using deduplication offsets instead of durations.
+An offset defines a point in the history of the ledger and is thus not affected by clock skews and network delays.
+Offsets are arguably less intuitive and require more effort by the application developer.
+We recommend the following strategy for using deduplication offsets:
+
+#. Choose a fresh command ID for the ledger change and the ``actAs`` parties, which (together with the application ID) determine the change ID.
+   Remember the command ID across application crashes.
+   (Analogous to :ref:`step 1 above <dedup-bounded-step-command-id>`)
+
+   .. _dedup-unbounded-step-dedup-offset:
+   
+#. Obtain a recent offset ``OFF0`` on the completion event stream and remember across crashes that you use ``OFF0`` with the chosen command ID.
+
+   - When you use the :ref:`Command Service <command-service>`, to obtain a recent offset, repeatedly submit a dummy command, e.g., a :ref:`Create-And-Exercise command <com.daml.ledger.api.v1.CreateAndExerciseCommand>` of some single-signatory template with the :ref:`Archive <function-da-internal-template-functions-archive-52202>` choice, until you get a successful response.
+     The response contains the :ref:`completion offset <com.daml.ledger.api.v1.SubmitAndWaitForTransactionIdResponse.completion_offset>`.
+
+   - When you use the :ref:`Command Completion Service <command-completion-service>`, ask for the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.
+
+   .. _dedup-unbounded-step-offset:
+
+#. When you use the :ref:`Command Completion Service <command-submission-service>`, obtain a recent offset on the completion stream ``OFF1``, say its current end.
+   (Analogous to :ref:`step 2 above <dedup-bounded-step-offset>`)
+
+   .. note::
+      Unless error handling retries from this step, you can use ``OFF1 = OFF0``.
```

Comment:
Error handling instructs the application developer to `retry from step 3`. In that case, you want to obtain a fresh offset `OFF1`. If you execute this step the first time, you can use `OFF0`, because if you request another ledger end, you'll most likely get `OFF0` again. I've rephrased this now.